src/Marten/Events/Daemon/HighWater/HighWaterDetector.cs

```diff
@@ -0,0 +1,169 @@
+            _tenant = tenant;
+
+            _findSafeSequence = new NpgsqlCommand($@"select min(seq_id) from {graph.DatabaseSchemaName}.mt_events where mt_events.timestamp >= :timestamp");
+            _safeTimestamp = _findSafeSequence.AddNamedParameter("timestamp", DateTimeOffset.MinValue);
+
+            _gapDetection = new NpgsqlCommand($@"
+select seq_id
+from   (select
+               seq_id,
+               lead(seq_id)
+               over (order by seq_id) as no
+        from
+               {graph.DatabaseSchemaName}.mt_events where seq_id > :start) ct
+where  no is not null
+  and    no - seq_id > 1
+LIMIT 1;
+select max(seq_id) from {graph.DatabaseSchemaName}.mt_events where seq_id > :start;
+".Trim());
+
+            _start = _gapDetection.AddNamedParameter("start", 0L);
+
+            _stateDetection = new NpgsqlCommand($@"
+select last_value from {graph.DatabaseSchemaName}.mt_events_sequence;
+select last_seq_id, last_updated, transaction_timestamp() as timestamp from {graph.DatabaseSchemaName}.mt_event_progression where name = '{ShardState.HighWaterMark}';
+".Trim());
+
+            _updateStatus =
+                new NpgsqlCommand($"select {graph.DatabaseSchemaName}.mt_mark_event_progression('{ShardState.HighWaterMark}', :seq);");
+            _newSeq = _updateStatus.AddNamedParameter("seq", 0L);
+
+        }
+
+        public async Task<HighWaterStatistics> DetectInSafeZone(DateTimeOffset safeTimestamp, CancellationToken token)
+        {
+            await using var conn = _tenant.OpenConnection();
+
+            var statistics = await loadCurrentStatistics(conn, token);
+
+            _safeTimestamp.Value = safeTimestamp;
+            using (var reader = await conn.ExecuteReaderAsync(_findSafeSequence, token))
+            {
+                if (await reader.ReadAsync(token))
+                {
+                    statistics.SafeStartMark = await reader.GetFieldValueAsync<long>(0, token);
+                }
+            }
+
+            await calculateHighWaterMark(token, statistics, conn);
+
+            return statistics;
+        }
+
+
+        public async Task<HighWaterStatistics> Detect(CancellationToken token)
+        {
+            await using var conn = _tenant.OpenConnection();
+
+            var statistics = await loadCurrentStatistics(conn, token);
+
+
+            await calculateHighWaterMark(token, statistics, conn);
+
+            return statistics;
+        }
+
+        private async Task calculateHighWaterMark(CancellationToken token, HighWaterStatistics statistics,
+            IManagedConnection conn)
+        {
+            // If the last high water mark is the same as the highest number
+            // assigned from the sequence, then the high water mark cannot
+            // have changed
+            if (statistics.LastMark == statistics.HighestSequence)
+            {
+                statistics.CurrentMark = statistics.LastMark;
+            }
+            else if (statistics.HighestSequence == 0)
```

Comment:
@jeremydmiller is this case when projection was resetted and is being rebuilt from scratch?

Reply:
It's there because it came up in the tests. The only thing that can possibly be is if the system is in its initial state